### ISSUE TITLE
Reranking + Deduplication

### ISSUE THREAD

**Issue by Quantakavin** [hyardim/Ambience-AI-1.5#43]
## Overview

Implement the sixth stage of the retrieval pipeline: reranking fused candidate results using a cross-encoder model for accurate relevance scoring, followed by deduplication to remove near-identical chunks from the final result set.

---

## File
`src/retrieval/rerank.py`

---

## Interface
```python
@dataclass
class RankedResult:
    chunk_id: str
    doc_id: str
    text: str
    rerank_score: float            # cross-encoder score (higher is better)
    rrf_score: float               # preserved from fusion stage
    vector_score: float | None     # preserved from fusion stage
    keyword_rank: float | None     # preserved from fusion stage
    metadata: dict[str, Any]

def rerank(
    query: str,
    results: list[FusedResult],
    top_k: int = 10,
    model_name: str = "cross-encoder/ms-marco-MiniLM-L-6-v2",
) -> list[RankedResult]:
    ...

def deduplicate(
    results: list[RankedResult],
    similarity_threshold: float = 0.85,
) -> list[RankedResult]:
    ...

def _jaccard_similarity(text_a: str, text_b: str) -> float:
    ...
```

---

### Cross-encoder scoring
- Load `cross-encoder/ms-marco-MiniLM-L-6-v2` from sentence-transformers
- For each candidate result, form a `(query, chunk_text)` pair
- Score all pairs in a single batched call for efficiency
- Sort results by `rerank_score` descending
- Return top `top_k` results

### Model loading
- Load model once on first call, cache in module-level variable
- Do not reload per query — same pattern as `embed.py` in ingestion

### Score normalisation
- Raw cross-encoder scores are logits — normalise to 0–1 using sigmoid:
```python
score = 1 / (1 + exp(-logit))
```
- Preserves relative ordering while making scores interpretable

### Input size
- Log a warning if input exceeds 50 results — reranking is expensive and should only receive top 20–30 candidates from fusion/filter stages
- Do not truncate silently — proceed with all results but warn

### Deduplication
- Compare all pairs of results using token-level Jaccard similarity:
```python
def _jaccard_similarity(text_a: str, text_b: str) -> float:
    tokens_a = set(text_a.lower().split())
    tokens_b = set(text_b.lower().split())
    if not tokens_a and not tokens_b:
        return 1.0
    return len(tokens_a & tokens_b) / len(tokens_a | tokens_b)
```
- If two results have Jaccard similarity above `similarity_threshold` (default `0.85`), keep the one with the higher `rerank_score` and drop the other
- Run deduplication after reranking so scores are available for comparison
- O(n²) comparison is acceptable for typical input sizes (≤30 results)
- Log how many results were dropped

---

### rerank()
- Empty input list → return `[]`, do not raise
- Model load failure → `RetrievalError(stage="RERANK", ...)`
- Scoring failure for a single pair → log warning, assign `rerank_score=0.0`, do not raise
- Input exceeds 50 results → log warning, proceed

### deduplicate()
- Empty input list → return `[]`, do not raise
- Invalid `similarity_threshold` (outside 0–1) → raise `ValueError`

---

## Logging
```
DEBUG | rerank.py | Reranking 24 candidates for query: "gout treatment options"
DEBUG | rerank.py | Reranking complete in 180ms
DEBUG | rerank.py | Top rerank score: 0.94, bottom: 0.12
DEBUG | rerank.py | Returning top 10 after reranking
DEBUG | rerank.py | Deduplication: dropped 2 near-duplicate results, 8 remaining
```

---

## Tests
`tests/retrieval/test_rerank.py`

All rerank tests mock the cross-encoder model — no real model loaded in unit tests.
Deduplication tests are pure functions — no mocking required.

### rerank()
| Test | Description |
|------|-------------|
| `test_returns_list_of_ranked_results` | Output is `list[RankedResult]` |
| `test_results_ordered_by_rerank_score_descending` | Highest score first |
| `test_top_k_limits_output` | Never returns more than `top_k` |
| `test_rerank_score_is_normalised_between_0_and_1` | Score in `[0, 1]` |
| `test_rrf_score_preserved_in_output` | `rrf_score` carried through |
| `test_vector_score_preserved_in_output` | `vector_score` carried through |
| `test_keyword_rank_preserved_in_output` | `keyword_rank` carried through |
| `test_empty_input_returns_empty_list` | Empty input returns `[]` |
| `test_model_load_failure_raises_retrieval_error` | Wrapped in `RetrievalError` |
| `test_single_pair_scoring_failure_assigns_zero_score` | Failed pair gets `0.0` |
| `test_large_input_logs_warning` | >50 results logs warning |
| `test_all_pairs_scored_in_single_batch` | Single `.predict()` call |
| `test_model_loaded_once_across_multiple_calls` | No reload between calls |

### deduplicate()
| Test | Description |
|------|-------------|
| `test_deduplicate_drops_near_duplicates` | High Jaccard similarity removed |
| `test_deduplicate_keeps_higher_scoring_result` | Higher `rerank_score` preserved |
| `test_deduplicate_preserves_unique_results` | Distinct chunks untouched |
| `test_deduplicate_empty_input_returns_empty` | Empty input returns `[]` |
| `test_jaccard_similarity_computed_correctly` | Known inputs produce correct score |
| `test_similarity_threshold_controls_aggressiveness` | Lower threshold drops more |
| `test_invalid_similarity_threshold_raises_value_error` | Outside 0–1 raises `ValueError` |

---

## Non-Functional Requirements
- Model loaded once, reused across calls
- Rerank latency target: <250ms for 20 candidates on CPU
- Deduplication latency target: O(n²) acceptable for ≤30 results
- 100% test coverage across both functions
- Model name configurable at call time for future upgrades
